ecc/bls12-377/fflonk/fflonk.go

```diff
@@ -0,0 +1,291 @@
+// Copyright 2020 Consensys Software Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Code generated by consensys/gnark-crypto DO NOT EDIT
+
+package fflonk
+
+import (
+	"errors"
+	"hash"
+	"math/big"
+
+	"github.com/consensys/gnark-crypto/ecc/bls12-377/fr"
+	"github.com/consensys/gnark-crypto/ecc/bls12-377/kzg"
+	"github.com/consensys/gnark-crypto/ecc/bls12-377/shplonk"
+)
+
+var (
+	ErrRootsOne                       = errors.New("Fr does not contain all the t-th roots of 1")
+	ErrNbPolynomialsNbPoints          = errors.New("the number of packs of polynomials should be the same as the number of pack of points")
+	ErrInonsistentFolding             = errors.New("the outer claimed values are not consistent with the shplonk proof")
+	ErrInconsistentNumberFoldedPoints = errors.New("the number of outer claimed values is inconsistent with the number of claimed values in the shplonk proof")
+)
+
+// Opening fflonk proof for opening a list of list of polynomials ((fʲᵢ)ᵢ)ⱼ where each
+// pack of polynomials (fʲᵢ)ᵢ (the pack is indexed by j) is opened on a powers of elements in
+// the set Sʲ (indexed by j), where the power is |(fʲᵢ)ᵢ|.
+//
+// implements io.ReaderFrom and io.WriterTo
+type OpeningProof struct {
+
+	// shplonk opening proof of the folded polynomials
+	SOpeningProof shplonk.OpeningProof
+
+	// ClaimedValues ClaimedValues[i][j] contains the values
+	// of fʲᵢ on Sⱼ^{|(fʲᵢ)ᵢ|}
+	ClaimedValues [][][]fr.Element
+}
+
+// CommitAndFold commits to a list of polynomial by intertwinning them like in the FFT, that is
+// returns ∑_{i<t}Pᵢ(Xᵗ)Xⁱ for t polynomials
+func CommitAndFold(p [][]fr.Element, pk kzg.ProvingKey, nbTasks ...int) (kzg.Digest, error) {
+	buf := Fold(p)
+	com, err := kzg.Commit(buf, pk, nbTasks...)
+	return com, err
+}
+
+// Fold returns p folded as in the fft, that is ∑_{i<t}Pᵢ(Xᵗ)Xⁱ. The
+func Fold(p [][]fr.Element) []fr.Element {
+
+	// we first pick the smallest divisor of r-1 bounding above len(p)
+	t := getNextDivisorRMinusOne(len(p))
+
+	sizeResult := 0
+	for i := range p {
+		if sizeResult < len(p[i]) {
+			sizeResult = len(p[i])
+		}
+	}
+	sizeResult = sizeResult*t + t - 1
```

Comment:
normally `sizeResult*t` is not sufficient, because the polynomials are intertwined. If you have 2 polynomials `f1` and `f2` of size n, the result is f1(X^t)+ X*f2(X^t), so then number of coefficients is bigger than `n*t`, it is `n*t +len(p)-1 <= n*t+t-1`, and in the ideal case where len(p)=t, then this upper bound is reached

Reply:
Ah, I see. I forgot the additional multiplication by powers of `X` in folding.

Reply:
Sorry what I wrote is wrong I think, if you have t polynomials fi of size n, their degree is (n-1) so fi(X^t) is of degree t(n-1). Since we shift the polynomials using f1(X^t)+Xf2(X^t)+.. + X^(t-1)ft(X^) the total degree is t(n-1)+t-1=nt-1, so sizeResult*t-1 (I think I interpreted sizeResult as the maximum degree)

Reply:
I added a comment to explain the degree